bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/IdentifierEncoder.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.jvm;
+
+import org.apache.commons.lang3.StringEscapeUtils;
+
+/**
+ * Identifier encoder to encode user defined identifiers with special characters.
+ *
+ * @since 2.0.0
+ */
+public class IdentifierEncoder {
+
+    private static final String CHAR_PREFIX = "$";
+    private static final String ESCAPE_PREFIX = "\\";
+    private static final String ENCODABLE_CHAR_SET = "\\.:;[]/<>$";
+    private static final String ENCODING_PATTERN = "\\$(\\d{4})";
+
+    private IdentifierEncoder() {
+    }
+
+    private static String encodeSpecialCharacters(String identifier) {
+        StringBuilder sb = new StringBuilder();
+        int index = 0;
+        while (index < identifier.length()) {
+            if (identifier.charAt(index) == '\\' && (index + 1 < identifier.length()) &&
+                    ENCODABLE_CHAR_SET.contains(Character.toString(identifier.charAt(index + 1)))) {
+                String unicodePoint = CHAR_PREFIX + String.format("%04d", (int) identifier.charAt(index + 1));
+                sb.append(unicodePoint);
+                index += 2;
+            } else {
+                sb.append(identifier.charAt(index));
+                index++;
+            }
+        }
+        return sb.toString();
+    }
+
+    public static String escapeSpecialCharacters(String identifier) {
+        String specialCharSet = "([$&+,:;=\\?@#|/' \\[\\}\\]<\\>.\"^*{}~`()%!-])";
+        return identifier.replaceAll("(?<!\\\\)(?:\\\\\\\\)*" + specialCharSet, "\\\\$1");
+    }
+
+    /**
+     * Encode the identifiers to avoid using jvm reserved characters.
+     *
+     * @param identifier identifier string
+     * @return encoded identifier
+     */
+    public static String encodeIdentifier(String identifier) {
+        if (identifier == null) {
+            return identifier;
+        }
+        if (identifier.contains(ESCAPE_PREFIX)) {
+            identifier = encodeSpecialCharacters(identifier);
+        } else {
+            identifier = identifier.replaceAll(ENCODING_PATTERN, "\\$#$1");
```

Comment:
replaceAll has a performance impact as it is internally compiled like the below 
```java
Pattern.compile(regex).matcher(str).replaceAll(repl)
```

So to get some improvement on performance on this encode method, shall we use the below approach ?

Have the Pattern.compile once (may be as a class level constant) and then call the match on it within the encode method? That way, it only compiles once

references - https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#replaceAll-java.lang.String-java.lang.String-

SO - https://stackoverflow.com/questions/1466959/string-replaceall-vs-matcher-replaceall-performance-differences

Reply:
Changed the `replaceAll` into `matcher.replaceAll()`